File: matlabcode/Csetminmax.m
% Function CSETMINMAX	resets value to within given limits
%	useful for ensuring that longitude is in correct range
%      [value] = Csetminmax(value,lowlim,uplim)
%	
%	input:
%		value - value to be limited
%		lolim = lower limit
%		uplimit = upper limit
%	output:
%		value - new value (between limits)
%
%	EXAMPLE:
%	a) force longitude 0<=longitude<360
%		longitude = -180;
%		[longitude] = Csetminmax(longitude,0,360)
%	b) force longitude -pi<=longitude<pi
%		longitude = 2*pi;
%		[longitude] = Csetminmax(longitude,-pi,pi)

function	[value] = Csetminmax(value,lowlim,uplim);

diff=uplim-lowlim;
if diff <=0
	disp('ERROR: upper limit must be higher than lower limit')
	return
end

ind = find(value>=uplim);
while isempty(ind)==0
	value(ind) = value(ind)-diff;
    ind = find(value>=uplim);
end

ind = find(value<lowlim);
while isempty(ind)==0
	value(ind) = value(ind)+diff;
	ind = find(value<lowlim);
end

return

File: ccode/main/csetminmax.cpp
#include "csetminmax.hpp"

#include <iostream>

using namespace std;

vector<double> adjustToLimits(vector<double> value, double lowerlim, double upperlim, double difference) {
	for(int i = 0; i < value.size(); i++) {
		while(value[i] >= upperlim || value[i] < lowerlim) {
			value[i] = (value[i] >= upperlim) ? value[i] - 2*difference : value[i] + 2*difference;
		}
	}
	return value;
}

double adjustToLimits(double value, double lowerlim, double upperlim, double difference) {
	while(value >= upperlim || value < lowerlim) {
		value = (value >= upperlim) ? value - difference : value + difference;
	}
	return value;
}

vector<double> cSetMinMax(vector<double> value, double lowerlim, double upperlim) {
	double difference = upperlim - lowerlim;
	if (difference <= 0) {
		cout << "Error: Upper limit must be higher than lower limit" << endl;
		vector<double> zeros(value.size());
		return zeros;
	}
	return adjustToLimits(value, lowerlim, upperlim, difference);
}

double cSetMinMax(double value, double lowerlim, double upperlim) {
	double difference = upperlim - lowerlim;
	if (difference <= 0) {
		cout << "Error: Upper limit must be higher than lower limit" << endl;
		return 0;
	}
	return adjustToLimits(value, lowerlim, upperlim, difference);
}